### PR TITLE
Integrate confirm entitlements endpoint

### PR DESCRIPTION
**Pull request by miasma13**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1206773931481624/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2541
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2325
What kind of version bump will this require?: Major

**Description**:
Integrate confirm entitlements endpoint for macOS App Store purchase with additional retries in case of failure.

**Steps to test this PR**:
See iOS and macOS PRs.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift [hunk 1]
```diff
@@ -108,14 +110,41 @@ public final class AppStorePurchaseFlow {
...
+    // swiftlint:enable cyclomatic_complexity
     @discardableResult
-    public static func completeSubscriptionPurchase() async -> Result<PurchaseUpdate, AppStorePurchaseFlow.Error> {
+    public static func completeSubscriptionPurchase(with transactionJWS: TransactionJWS) async -> Result<PurchaseUpdate, AppStorePurchaseFlow.Error> {
         os_log(.info, log: .subscription, "[AppStorePurchaseFlow] completeSubscriptionPurchase")
 
-        let result = await AccountManager.checkForEntitlements(wait: 2.0, retry: 20)
+        guard let accessToken = AccountManager().accessToken else { return .failure(.missingEntitlements) }
+
+        let result = await callWithRetries(wait: 0.5, retry: 3) {
```
Comment (afterxleep) [inline Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift:121]: Isn't this wait too short?  Why only 3 retries?

File: Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift [hunk 2]
```diff
@@ -108,14 +110,41 @@ public final class AppStorePurchaseFlow {
...
+                return true
+            case .failure:
+                return false
+            }
+        }
 
         return result ? .success(PurchaseUpdate(type: "completed")) : .failure(.missingEntitlements)
     }
+
+    private static func callWithRetries(wait waitTime: Double, retry retryCount: Int, conditionToCheck: () async -> Bool) async -> Bool {
```
Comment (afterxleep) [inline Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift:133]: WDYT about moving the retry logic to `executeAPICall`in the APIService?   That way, we could easily implement retry in every Backend call.

File: Sources/Subscription/Services/Model/Subscription.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
+//
+//  Subscription.swift
```
Comment (afterxleep) [inline Sources/Subscription/Services/Model/Subscription.swift:2]: 👍

File: Sources/Subscription/Services/SubscriptionService.swift [hunk 4]
```diff
@@ -101,4 +82,19 @@ public struct SubscriptionService: APIService {
...
     }
 
+    // MARK: -
+
+    public static func confirmPurchase(accessToken: String, signature: String) async -> Result<ConfirmPurchaseResponse, APIServiceError> {
+        let headers = makeAuthorizationHeader(for: accessToken)
+        let bodyDict = ["signedTransactionInfo": signature]
+
+        guard let bodyData = try? JSONEncoder().encode(bodyDict) else { return .failure(.encodingError) }
+        return await executeAPICall(method: "POST", endpoint: "purchase/confirm/apple", headers: headers, body: bodyData)
```
Comment (afterxleep) [inline Sources/Subscription/Services/SubscriptionService.swift:92]: Maybe move the strings in this method to constants?

**Review by afterxleep — APPROVED** [review]
Looking good, a couple of suggestions.   Up to you to implement

File: Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift (hunk 1 above)
Comment (miasma13) [reply Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift:121]: What are your suggestions?

File: Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift (hunk 2 above)
Comment (miasma13) [reply Sources/Subscription/Flows/AppStore/AppStorePurchaseFlow.swift:133]: I think it is worth considering, though I would defer it until we have a need for other calls to be made with retires.

File: Sources/Subscription/Services/SubscriptionService.swift (hunk 4 above)
Comment (miasma13) [reply Sources/Subscription/Services/SubscriptionService.swift:92]: Makes sense but I am considering move the implementation to use `Networking` package.

**Pull request merged**